l1-contracts/contracts/bridge/asset-tracker/L1AssetTracker.sol

```diff
@@ -0,0 +1,372 @@
+    }
+
+    /// @notice This function is used to migrate the token balance from the NTV to the AssetTracker for V30 upgrade.
+    /// @param _chainId The chain id of the chain to migrate the token balance for.
+    /// @param _assetId The asset id of the token to migrate the token balance for.
+    function migrateTokenBalanceFromNTVV30(uint256 _chainId, bytes32 _assetId) external {
+        IL1NativeTokenVault l1NTV = IL1NativeTokenVault(address(NATIVE_TOKEN_VAULT));
+        uint256 migratedBalance;
+        if (_chainId != block.chainid) {
+            migratedBalance = l1NTV.migrateTokenBalanceToAssetTracker(_chainId, _assetId);
+        } else {
+            address tokenAddress = NATIVE_TOKEN_VAULT.tokenAddress(_assetId);
+            migratedBalance = IERC20(tokenAddress).totalSupply();
+            require(chainBalance[block.chainid][_assetId] == 0, ChainBalanceNotZero());
+        }
+        /// Note it might be the case that the tokenOriginChainId and the specified _chainId are both L1,
+        /// in this case the chainBalance[L1_CHAIN_ID][_assetId] is set to uint256.max if it was not already.
+        uint256 originChainId = NATIVE_TOKEN_VAULT.originChainId(_assetId);
+        /// kl todo can it be the case that we set chainBalance to uint256.max twice.
+        if (chainBalance[originChainId][_assetId] == 0) {
+            chainBalance[originChainId][_assetId] = MAX_TOKEN_BALANCE - migratedBalance;
+        } else {
+            chainBalance[originChainId][_assetId] -= migratedBalance;
+        }
+        chainBalance[_chainId][_assetId] += migratedBalance;
+    }
+
+    /// @notice This is used to register L2NativeTokens when the chain is settling on GW.
+    /// @notice It is needed since withdrawals are blocked until the token balance is migrated to GW, and it needs to be registered before migration.
+    function registerL2NativeToken(uint256 _l2ChainId, address _l2NativeToken) external {
+        uint256 settlementLayer = BRIDGE_HUB.settlementLayer(_l2ChainId);
+        require(settlementLayer != block.chainid, InvalidSettlementLayer());
+
+        bytes32 assetId = DataEncoding.encodeNTVAssetId(_l2ChainId, _l2NativeToken);
+
+        require(NATIVE_TOKEN_VAULT.tokenAddress(assetId) == address(0), InvalidTokenAddress());
+
+        chainBalance[settlementLayer][assetId] = MAX_TOKEN_BALANCE;
+
+        _sendConfirmationToChains(
+            settlementLayer,
+            ConfirmBalanceMigrationData({
+                version: TOKEN_BALANCE_MIGRATION_DATA_VERSION,
+                isL1ToGateway: true,
+                chainId: settlementLayer,
+                assetId: assetId,
+                migrationNumber: _getChainMigrationNumber(_l2ChainId),
+                amount: MAX_TOKEN_BALANCE
+            })
+        );
+    }
+
+    function registerNewToken(bytes32 _assetId, uint256 _originChainId) public override onlyNativeTokenVault {
+        _assignMaxChainBalance(_originChainId, _assetId);
```

Comment:
what prevents the following scenario:

- Chain made a withdrawal (unfinalized)
- Then it moved to GW
- it calls `registerL2NativeToken` (assigns 2^256-1 to SL)
- it finalizes the withdrawal and so `_assignMaxChainBalance` is called and assigns 2^256-1 balance to the chain. 

The total sum of chain balances is now. >2^256-1